### ISSUE TITLE
Query GIS for Sale Tax Rate by Address

### ISSUE THREAD

**Issue by uncheckederror** [uncheckederror/WASalesTaxLookup#2]
Maybe we can use these map services to skip around the whole CSV importing process?

https://webgis2.dor.wa.gov/arcgis/rest/services/Base/WADOR_Base_Labels_ST/MapServer

**Comment by dewrightca** [comment]
Don't tag anything to webgis2; go to the parent LB url of https://webgis.dor.wa.gov.

The route you would want to go is thus; 
1) Geocode using https://webgis.dor.wa.gov/arcgis/rest/services/Locators/WA_Address/GeocodeServer
2) Take the coordinate from 1) and query it against https://webgis.dor.wa.gov/arcgis/rest/services/Programs/WADOR_SalesTax/MapServer/4 to return the LOCCODE value
3) Take the LOCCODE from 2) and retrieve that from the Rates table you are already harvesting from the exist CSV files.

**Comment by uncheckederror** [comment]
Okay, awesome! Thanks @dewrightca this is just the info I needed to make this work. 🎉

I'm writing the specific workflow steps here so I don't forget them:

- Geocode the address and get some JSON back by calling https://webgis.dor.wa.gov/arcgis/rest/services/Locators/WA_Address/GeocodeServer/findAddressCandidates?Address=1622+S+Graham+St&Address2=&Address3=&Neighborhood=&City=Seattle&Subregion=WA&Region=&Postal=&PostalExt=&CountryCode=&SingleLine=&outFields=&maxLocations=&matchOutOfRange=true&langCode=&locationType=&sourceCountry=&category=&location=&searchExtent=&outSR=&magicKey=&preferredLabelValues=&f=pjson
- Take the x, y coordinates from the location field in the top result then query it against this endpoint https://webgis.dor.wa.gov/arcgis/rest/services/Programs/WADOR_SalesTax/MapServer/4/query?where=&text=&objectIds=&time=&timeRelation=esriTimeRelationOverlaps&geometry=-122.311388543841%2C+47.547320805635&geometryType=esriGeometryPoint&inSR=4326&spatialRel=esriSpatialRelIntersects&distance=&units=esriSRUnit_Foot&relationParam=&outFields=LOCCODE&returnGeometry=false&returnTrueCurves=false&maxAllowableOffset=&geometryPrecision=&outSR=&havingClause=&returnIdsOnly=false&returnCountOnly=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&returnZ=false&returnM=false&gdbVersion=&historicMoment=&returnDistinctValues=false&resultOffset=&resultRecordCount=&returnExtentOnly=false&sqlFormat=none&datumTransformation=&parameterValues=&rangeValues=&quantizationParameters=&featureEncoding=esriDefault&f=pjson
- Take the LOCCODE value from the result and find the matching value the Code column of the Rates_25Q1.csv file.
- Format the Rate into XML (or JSON) and then return it to the user.

This is pretty doable. I'm updating this project to use the new [Minimal API's style/framework](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-9.0) so this change will arrive along with that work.

This method does take a direct dependency on the https://webgis.dor.wa.gov/ infrastructure. But I imagine it's pretty stable, perhaps I could maintain some sort of fallback to working purely off the CSV exports incase there's an outage or something. I'll have to think about that more.

**Comment by dewrightca** [comment]
> Okay, awesome! Thanks [@dewrightca](https://github.com/dewrightca) this is just the info I needed to make this work. 🎉
> 
> I'm writing the specific workflow steps here so I don't forget them:
> 
> * Geocode the address and get some JSON back by calling https://webgis.dor.wa.gov/arcgis/rest/services/Locators/WA_Address/GeocodeServer/findAddressCandidates?Address=1622+S+Graham+St&Address2=&Address3=&Neighborhood=&City=Seattle&Subregion=WA&Region=&Postal=&PostalExt=&CountryCode=&SingleLine=&outFields=&maxLocations=&matchOutOfRange=true&langCode=&locationType=&sourceCountry=&category=&location=&searchExtent=&outSR=&magicKey=&preferredLabelValues=&f=pjson
> * Take the x, y coordinates from the location field in the top result then query it against this endpoint https://webgis.dor.wa.gov/arcgis/rest/services/Programs/WADOR_SalesTax/MapServer/4/query?where=&text=&objectIds=&time=&timeRelation=esriTimeRelationOverlaps&geometry=-122.311388543841%2C+47.547320805635&geometryType=esriGeometryPoint&inSR=4326&spatialRel=esriSpatialRelIntersects&distance=&units=esriSRUnit_Foot&relationParam=&outFields=LOCCODE&returnGeometry=false&returnTrueCurves=false&maxAllowableOffset=&geometryPrecision=&outSR=&havingClause=&returnIdsOnly=false&returnCountOnly=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&returnZ=false&returnM=false&gdbVersion=&historicMoment=&returnDistinctValues=false&resultOffset=&resultRecordCount=&returnExtentOnly=false&sqlFormat=none&datumTransformation=&parameterValues=&rangeValues=&quantizationParameters=&featureEncoding=esriDefault&f=pjson
> * Take the LOCCODE value from the result and find the matching value the Code column of the Rates_25Q1.csv file.
> * Format the Rate into XML (or JSON) and then return it to the user.
> 
> This is pretty doable. I'm updating this project to use the new [Minimal API's style/framework](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-9.0) so this change will arrive along with that work.
> 
> This method does take a direct dependency on the https://webgis.dor.wa.gov/ infrastructure. But I imagine it's pretty stable, perhaps I could maintain some sort of fallback to working purely off the CSV exports incase there's an outage or something. I'll have to think about that more.

Yes that is the same guidance we provide to other organizations using the older version of the service API (https://dor.wa.gov/washington-sales-tax-rate-library-source-code) ; to go ahead and build their service to either call our WebAPI directly (https://dor.wa.gov/wa-sales-tax-rate-lookup-url-interface) and then use the local service as a fallback.

**Comment by uncheckederror** [comment]
Yep, that makes perfect sense. These changes are live here: https://wataxlookup.acceleratenetworks.com/scalar/v1

I have updated this project to follow that workflow. The /GetTaxRate endpoint now uses this GIS-based lookup method, while the /AddressRates.aspx maintains the original lookup method and formatting as (https://dor.wa.gov/wa-sales-tax-rate-lookup-url-interface).

Thanks again for your help!
